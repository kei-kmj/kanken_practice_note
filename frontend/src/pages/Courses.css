.coursesTitle {
    position: relative;
    padding: 1rem 2rem;
}


