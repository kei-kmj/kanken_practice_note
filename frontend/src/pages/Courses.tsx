import * as React from 'react'
import { useRef, useState } from 'react'
import { useNavigate } from 'react-router-dom'
import { useFetchApi } from '../hooks/useFetchAPI'
import { Header} from '../components/shared/Header'
import { Logo } from '../components/shared/Logo'
import { LevelPanel } from '../components/courses/LevelPanel'
import { CategoryPanel } from '../components/courses/CategoryPanel'
import { QuizStartPanel } from '../components/courses/QuizStartPanel'
import { BackToTop } from '../components/shared/BackToTop'
import { Footer } from '../components/shared/Footer'
import { LOCAL_STORAGE_DATA } from '../hooks/useLocalStrage'
import { CATEGORY, LEVEL } from '../constants'
import '../App.css'

export const Courses:React.FC = () => {
  const { fetchRequest } = useFetchApi()
  const navigate = useNavigate()
  const [level, setLevel] = useState<string>('')
  const [category, setCategory] = useState<string>('')

  const ref = useRef<HTMLDivElement>(null!)
  const scrollToStartButton = () => {
    ref.current.scrollIntoView({
      behavior: "smooth",
    });
  }

  const selectCategory = (category: string) => {
    setCategory(category)
    scrollToStartButton()
  }

  const startQuiz = async (): Promise<void> => {
    const {quiz} = await fetchRequest(level, category, '10')
    // console.log(quiz)

    if (quiz.length === 0) {
      alert('出題できる問題がありません')
      return
    }
    navigate('/quiz', {state: {quiz}})
  }

  const repeatQuiz = () => {
    const quiz = []
    for (const key in localStorage) {
      if (!key.startsWith(LOCAL_STORAGE_DATA.KEY)) {
        continue
      } else if (level !== JSON.parse(localStorage.getItem(key) as string).level.toString()) {
        continue
      } else if (category === CATEGORY.ALL ||
        category === JSON.parse(localStorage.getItem(key) as string).category_id.toString()) {
        quiz.push(JSON.parse(localStorage.getItem(key) as string))
      }
    }
    console.log(quiz)
    if (quiz.length === 0) {
      alert('登録されている問題がありません')
      return
    }
    navigate('/quiz', {state: {quiz}})
  }
  return (<>
    <Header pageTitle="コース選択" description="問題の級と分野を選択します"/>
    <div className="flex flex-col min-h-screen">
      <main className="flex flex-col flex-1 top-0">
        <Logo/>
        <LevelPanel
          onClickSemiFirst={() => setLevel(LEVEL.SEMI_FIRST)}
          onClickFirst={() => setLevel(LEVEL.FIRST)}/>
        {level &&
        <div>
          <CategoryPanel
            onClickAll={() => selectCategory(CATEGORY.ALL)}
            onClickReading={() => selectCategory(CATEGORY.READING)}
            onClickExtraReading={() => selectCategory(CATEGORY.EXTRA_READING)}
            onClickDictation={() => selectCategory(CATEGORY.DICTATION)}
            onClickDictationIdiom={() => selectCategory(CATEGORY.DICTATION_IDIOM)}
            onClickMeaningIdiom={() => selectCategory(CATEGORY.MEANING_IDIOM)}
            onClickAntonymSynonym={() => selectCategory(CATEGORY.ANTONYM_SYNONYM)}
            onClickProverb={() => selectCategory(CATEGORY.PROVERB)}/>

          {category && <QuizStartPanel onClickStart={startQuiz} onClickRepeat={repeatQuiz}/>}
        </div>}
        <div ref={ref}/>
        <BackToTop/>
      </main>
      <Footer/>
    </div>
  </>)
}
