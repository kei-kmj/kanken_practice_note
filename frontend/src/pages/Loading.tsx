import * as React from 'react'

export const Loading:React.FC = () => {
  return (
    <div className="flex flex-col min-h-screen">
      <div className="flex justify-center">
        <div className="animate-spin h-10 w-10 border-4 border-accent rounded-full border-t-transparent"/><p className="text-accent text-xl">　...しばらくお待ちください</p>
      </div>
    </div>

  )
}
