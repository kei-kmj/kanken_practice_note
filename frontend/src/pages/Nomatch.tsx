import * as React from 'react'
import { BackToTop } from '../components/shared/BackToTop'

export const Nomatch:React.FC = () => {
  return (
    <div className="flex flex-col min-h-screen"><h2>404</h2>
      <h3>このページは存在しません</h3>
      <BackToTop/>
    </div>

  )
}
