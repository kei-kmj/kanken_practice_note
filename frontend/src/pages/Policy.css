.policy {
    background-color: #cc7000;
    height: 70px;
    text-align: center;
    color: #eaf4fc;
    padding: 0.5em 0.3em 0.8em 0.5em;
    border-radius: 8px;
}