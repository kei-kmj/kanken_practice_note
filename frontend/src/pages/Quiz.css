.modal {
    width: 380px;
}