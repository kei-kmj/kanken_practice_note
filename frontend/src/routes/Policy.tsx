import * as React from 'react'
import { Link } from 'react-router-dom'
import { Footer } from '../components/shared/Footer'
import './policy.css'

export const Policy:React.FC = () => {
  return (
    <>
      <div>
        <h1 className="policy text-4xl m-4 flex justify-center">アプリケーションポリシー</h1>
        <div>
          <h2 className="badge badge-lg badge-info m-4 mr-0">1</h2><span className="text-2xl m-5">ご利用にあたって</span>
          <p className="ml-5">当アプリの利用条件を予告なく変更することがあります。最新の情報をご確認いただきますようお願いいたします。</p>
        </div>
        <div>
          <h2 className="badge badge-lg badge-info m-4 mt-10 mr-0">2</h2><span
          className="text-2xl m-5">掲載情報について</span>
          <p className="ml-5">コンテンツ作成においては、細心の注意を払っておりますが、内容の正確性を保証するものではありません。</p>
          <p className="ml-5">正解とした選択肢以外にも別解がある場合があります。</p>
          <p className="ml-5">予告なしにコンテンツを変更、削除すること並びにアプリ自体の運営を中止することがあります。</p></div>

        <div>
          <h2 className="badge badge-lg badge-info m-4 mt-10 mr-0">3</h2><span
          className="text-2xl m-5">免責事項について</span>
          <p className="ml-5">漢字検定の合格を保証するものではありません。</p>
          <p className="ml-5">当アプリの利用はご利用者様の責任において行われるものとします。</p>
          <p className="ml-5">当アプリ利用中に発生した損失・損害に対し、いかなる責任も負いません。</p>
        </div>

        <div>
          <h2 className="badge badge-lg badge-info m-4 mt-10 mr-0">4</h2><span
          className="text-2xl m-5">使用技術について</span>

          <h3 className="text-xl mb-2 ml-5">ローカルストレージ</h3>
        </div>
        <div>
          <p className="ml-5">ローカルストレージとは、ウェブサイトが個のデバイス上でデータを保存、または取得できるようにする業界標準の技術です。</p>
          <p className="ml-5">当アプリでは出題する問題を選択するためのデータをローカルストレージに保存することがあります。</p>
          <p className="ml-5">当アプリの操作で個人情報をローカルストレージに保存することは一切ありません。</p>
          <p className="ml-5">ブラウザの設定によりLocalstorageの設定を変更し、機能を無効にすることができますが、当アプリの一部の機能がご利用いただけなくなることがあります。</p>
          <p className="ml-5">Localstorageに保存したデータはトップページから削除出来ますが、一度削除すると元に戻せません。</p>
          <div>
            <h3 className="text-xl mt-5 mb-2 ml-5">JavaScript</h3>
            <p className="ml-5">当アプリは、より快適にご利用いただくため、JavaScriptを使用しています。 ご使用のブラウザ設定においてJavaScriptをオン (有効)
              にされていない場合に、正しく表示されない又は操作ができない場合があります。</p>
          </div>

        </div>
        <div className=" text-l grid grid-flow-col gap-4 link m-5 flex justify-end">
          <Link to="/">トップ画面に戻る</Link>
        </div>
      </div>
      <Footer/>
    </>)
}
