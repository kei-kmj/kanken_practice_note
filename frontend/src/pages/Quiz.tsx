import * as React from 'react'
import { useState } from 'react'
import { useNavigate, useLocation } from 'react-router-dom'
import { LOCAL_STORAGE_DATA, saveJSON, removeJSON } from '../hooks/useLocalStrage'
import { Header } from '../components/shared/Header'
import { Logo } from '../components/shared/Logo'
import { Question } from '../components/quiz/Question'
import { Alternatives } from '../components/quiz/Alternatives'
import { DisplayCorrect } from '../components/quiz/DisplayCorrect'
import { DisplayIncorrect } from '../components/quiz/DisplayIncorrect'
import { Note } from '../components/quiz/Note'
import { RepeatButton } from '../components/quiz/RepeatButton'
import { NextQuiz } from '../components/quiz/NextQuiz'
import { AbortConfirm } from '../components/quiz/AbortConfirm'
import { EndOfQuiz } from '../components/quiz/EndOfQuiz'
import { Score } from '../components/quiz/Score'
import { Footer } from '../components/shared/Footer'
import { QuestionType } from '../types/quizTypes'
import './Quiz.css'

export const Quiz:React.FC = () => {
  const navigate = useNavigate()
  const location = useLocation()

  const data = location.state as { quiz:QuestionType }
  // console.log(data)
  const [currentQuiz, setCurrentQuiz] = useState<number>(0)
  const [score, setScore] = useState<number>(0)

  const correctAnswer = data.quiz[currentQuiz].answers.filter((answer: { correctness: boolean }) => {
    if (answer.correctness) {
      return answer
    } else return null
  })

  const finishQuiz = () => {
    navigate('/courses')
  }

  const addScore = (answer: { correctness: boolean }) => {
    if (answer.correctness) {
      setScore(score + 1)
    }
  }

  const handleToggle = (e: { target: { checked: boolean } }) => {
    if (e.target.checked) {
      saveJSON(`${LOCAL_STORAGE_DATA.KEY}${data.quiz[currentQuiz].id}`, data.quiz[currentQuiz])
    } else {
      removeJSON(`${LOCAL_STORAGE_DATA.KEY}${data.quiz[currentQuiz].id}`)
    }
  }

  return (<>
    <Header pageTitle="問題です" description="問題を出題します" title="漢検練習帳"/>
    <div className="flex flex-col min-h-screen">
      <div className="flex justify-left">
        <Logo />
      </div>
      <main className="flex-grow">
        <Question data={data} currentQuiz={currentQuiz}/>
        {data.quiz[currentQuiz].answers.map((answer: { id: string; correctness: boolean; answer: string }) =>
          <div className="flex justify-center mt-2" key={answer.id}>

            <Alternatives answer={answer} onClick={() => addScore(answer)}/>

            <input type="checkbox" id={answer.id} className="modal-toggle"/>
            <div className="modal">
              <div className="modal-box bg-white">
                <div>
                  <div className="flex justify-center">
                    {answer.correctness.toString() === 'true' ? <DisplayCorrect/> : <DisplayIncorrect/>}
                  </div>
                  <p className="text-2xl flex justify-center m-5 ">答え：{correctAnswer[0].answer}</p>

                  <Note data={data} currentQuiz={currentQuiz}/>

                  <RepeatButton data={data} currentQuiz={currentQuiz} onClick={(e) => handleToggle(e)}/>

                  {currentQuiz === data.quiz.length - 1 ? <Score score={score} data={data}/> : null}

                  {currentQuiz < data.quiz.length - 1
                    ? <NextQuiz answer={answer} onClick={() => {
                      setCurrentQuiz(currentQuiz + 1)
                    }}/>
                    : <EndOfQuiz answer={answer} onClick={finishQuiz}/>}

                  {currentQuiz < data.quiz.length - 1
                    ? <AbortConfirm answer={answer} onClick={finishQuiz}/>
                    : null}
                </div>
              </div>
            </div>
          </div>)}
      </main>
      <Footer/>
    </div>
  </>)
}