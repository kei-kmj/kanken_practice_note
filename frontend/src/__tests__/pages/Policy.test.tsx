import { describe, test } from 'vitest'
import { Policy } from '../../pages/Policy'
import { BrowserRouter as Router } from 'react-router-dom';
import { render, screen } from '@testing-library/react'

describe('スナップショット', () => {
  test('スナップショットが同一', () => {
    const {container} = render(<Router><Policy /></Router>)
    expect(container.firstChild).toMatchSnapshot()
  })

  test('すべての要素が正常にレンダリングされている', () => {
    render(<Router><Policy /></Router>)
    expect(screen.getAllByRole("heading")).toBeTruthy()
    expect(screen.getAllByRole("generic")).toBeTruthy()
    expect(screen.getAllByRole("link")).toBeTruthy()

    expect(screen.getByText("「漢検」「漢字検定」は")).toBeInTheDocument()
    expect(screen.getByText("公益財団法人 日本漢字能力検定協会の登録商標です")).toBeInTheDocument()

  })
})