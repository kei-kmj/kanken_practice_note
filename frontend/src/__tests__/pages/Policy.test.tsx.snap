// Vitest Snapshot v1

exports[`スナップショット > スナップショットが同一 1`] = `
<div>
  <h1
    class="policyTitle text-2xl mb-5"
  >
    アプリケーションポリシー
  </h1>
  <div
    class="policy"
  >
    <div>
      <h2
        class="badge badge-lg badge-warning m-4 mr-0"
      >
        1
      </h2>
      <span
        class="text-2xl m-5"
      >
        ご利用にあたって
      </span>
      <p
        class="ml-5"
      >
        当アプリの利用条件を予告なく変更することがあります。最新の情報をご確認いただきますようお願いいたします。
      </p>
    </div>
    <div>
      <h2
        class="badge badge-lg badge-warning m-4 mt-10 mr-0"
      >
        2
      </h2>
      <span
        class="text-2xl m-5"
      >
        掲載情報について
      </span>
      <p
        class="ml-5"
      >
        コンテンツ作成においては、細心の注意を払っておりますが、内容の正確性を保証するものではありません。
      </p>
      <p
        class="ml-5"
      >
        正解とした選択肢以外にも別解がある場合があります。
      </p>
      <p
        class="ml-5"
      >
        予告なしにコンテンツを変更、削除すること並びにアプリ自体の運営を中止することがあります。
      </p>
    </div>
    <div>
      <h2
        class="badge badge-lg badge-warning m-4 mt-10 mr-0"
      >
        3
      </h2>
      <span
        class="text-2xl m-5"
      >
        免責事項について
      </span>
      <p
        class="ml-5"
      >
        漢字検定の合格を保証するものではありません。
      </p>
      <p
        class="ml-5"
      >
        当アプリの利用はご利用者様の責任において行われるものとします。
      </p>
      <p
        class="ml-5"
      >
        当アプリ利用中に発生した損失・損害に対し、いかなる責任も負いません。
      </p>
    </div>
    <div>
      <h2
        class="badge badge-lg badge-warning m-4 mt-10 mr-0"
      >
        4
      </h2>
      <span
        class="text-2xl m-5"
      >
        使用技術について
      </span>
      <h3
        class="text-xl mb-2 ml-5"
      >
        ローカルストレージ
      </h3>
    </div>
    <div>
      <p
        class="ml-5"
      >
        ローカルストレージとは、ウェブサイトが個のデバイス上でデータを保存、または取得できるようにする業界標準の技術です。
      </p>
      <p
        class="ml-5"
      >
        当アプリでは出題する問題を選択するためのデータをローカルストレージに保存することがあります。
      </p>
      <p
        class="ml-5"
      >
        当アプリの操作で個人情報をローカルストレージに保存することは一切ありません。
      </p>
      <p
        class="ml-5"
      >
        ブラウザの設定によりローカルストレージの設定を変更し、機能を無効にすることができますが、当アプリの一部の機能がご利用いただけなくなることがあります。
      </p>
      <p
        class="ml-5"
      >
        ローカルストレージに保存したデータはトップページから削除出来ますが、一度削除すると元に戻せません。
      </p>
      <div>
        <h3
          class="text-xl mt-5 mb-2 ml-5"
        >
          JavaScript
        </h3>
        <p
          class="ml-5"
        >
          当アプリは、より快適にご利用いただくため、JavaScriptを使用しています。 ご使用のブラウザ設定においてJavaScriptをオン (有効) にされていない場合に、正しく表示されない又は操作ができない場合があります。
        </p>
      </div>
    </div>
    <div
      class=" text-l grid grid-flow-col gap-4 link m-5 flex justify-end"
    >
      <a
        href="/"
      >
        トップ画面に戻る
      </a>
    </div>
  </div>
</div>
`;
